query-engine/core/src/schema_builder/input_types/objects/create_objects.rs

```diff
@@ -0,0 +1,286 @@
+                        // created at a later point, to avoid recursing too deep
+                        // (that has caused stack overflows on large schemas in
+                        // the past).
+                        ctx.nested_create_inputs_queue
+                            .push((Arc::clone(&input_object), Arc::clone(&rf)));
+
+                        Arc::downgrade(&input_object)
+                    }
+                };
+
+                let all_required_scalar_fields_have_defaults = rf
+                    .linking_fields()
+                    .scalar_fields()
+                    .all(|scalar_field| scalar_field.default_value.is_some());
+
+                let input_field = input_field(rf.name.clone(), InputType::object(input_object), None);
+
+                if rf.is_required && !all_required_scalar_fields_have_defaults {
+                    Some(input_field)
+                } else {
+                    Some(input_field.optional())
+                }
+            }
+        })
+        .collect()
+}
+
+fn field_should_be_kept_for_checked_create_input_type(field: &ScalarFieldRef) -> bool {
+    !field.is_auto_generated_int_id
+}
+
+/// Builds the create input type (<x>UncheckedCreateInput / <x>UncheckedCreateWithout<y>Input)
+/// Also valid for nested inputs. A nested input is constructed if the `parent_field` is provided.
+/// "Unchecked" input refers to allowing to write _all_ scalars on a model directly, which can
+/// lead to unintended data integrity violations if used incorrectly.
+fn unchecked_create_input_type(
+    ctx: &mut BuilderContext,
+    model: &ModelRef,
+    parent_field: Option<&RelationFieldRef>,
+) -> InputObjectTypeWeakRef {
+    // We allow creation from both sides of the relation - which would lead to an endless loop of input types
+    // if we would allow to create the parent from a child create that is already a nested create.
+    // To solve it, we remove the parent relation from the input ("Without<Parent>").
+    let name = match parent_field.map(|pf| pf.related_field()) {
+        Some(ref f) => format!(
+            "{}UncheckedCreateWithout{}Input",
+            model.name,
+            capitalize(f.name.as_str())
+        ),
+        _ => format!("{}UncheckedCreateInput", model.name),
+    };
+
+    return_cached_input!(ctx, &name);
+
+    let input_object = Arc::new(init_input_object_type(name.clone()));
+    ctx.cache_input_type(name, input_object.clone());
+
+    let linking_fields = if let Some(parent_field) = parent_field {
+        let child_field = parent_field.related_field();
+        if child_field.is_inlined_on_enclosing_model() {
+            child_field.linking_fields().scalar_fields().collect()
+        } else {
+            vec![]
+        }
+    } else {
+        vec![]
+    };
+
+    let scalar_fields: Vec<ScalarFieldRef> = model
+        .fields()
+        .scalar()
+        .into_iter()
+        .filter(|sf| !linking_fields.contains(sf))
+        .collect();
```

Comment:
I think this omits the scalar fields that are part of the relation that is traversed currently, right? This could be made clear through a comment.

Reply:
Linking fields is consistent vocabulary throughout the core.